src/alto-ui/Form/TagInput/TagInput.js

```diff
@@ -0,0 +1,258 @@
+import React, { Fragment, useState, useRef, useEffect } from 'react';
+import PropTypes from 'prop-types';
+
+import { bemClass } from '../../helpers/bem';
+
+import useEventListener from '../../hooks/useEventListener';
+import useFocusInOut from '../../hooks/useFocusInOut';
+import Tag from '../../Tag';
+
+import './TagInput.scss';
+
+const sliceArr = (xs, i, j) =>
+  xs.reduce((acc, x, k) => {
+    if (k >= i && k <= j && i < j) return [...acc, x];
+    if (k <= i && k >= j && i > j) return [x, ...acc];
+    return acc;
+  }, []);
+
+const TagInput = React.forwardRef(
+  (
+    {
+      className,
+      value: tags = [],
+      onChange,
+      rounded,
+      getTagValue,
+      getTagTitle,
+      disabled,
+      ...props
+    },
+    ref
+  ) => {
+    const [value, setValue] = useState('');
+    const [position, setPosition] = useState(undefined);
+    const [selection, setSelection] = useState([]);
+
+    const inputRef = useRef();
+    const tagInputValueRef = useRef();
+    const mainRef = useRef();
+    const tagInputRef = ref || mainRef;
+    const isActive = useFocusInOut(null, tagInputRef);
+
+    const isTagEqual = a => b => getTagValue(a) === getTagValue(b);
+    const isTagDiff = a => b => !isTagEqual(a)(b);
+    const isTagActive = tag => selection.some(isTagEqual(tag));
+
+    const availableTags = props.availableTags(value);
+    const remainingTags = (availableTags || []).filter(t => !tags.some(isTagEqual(t)));
+    const match =
+      props.match ||
+      (v =>
+        availableTags
+          ? remainingTags.some(t => getTagValue(t) === v)
+          : !tags.some(t => getTagValue(t) === v));
+
+    const addToSelection = (...xs) =>
+      setSelection([...selection, ...xs].filter((t, i, ts) => i === ts.findIndex(isTagEqual(t))));
+    const removeFromSelection = tag => setSelection(selection.filter(isTagDiff(tag)));
+
+    const renderTag =
+      props.renderTag || ((tag, tagProps) => <Tag {...tagProps}>{getTagTitle(tag)}</Tag>);
+
+    const onNewTag = props.onNewTag || ((tag, add) => add(tag));
+
+    const onRemoveTag = props.onRemoveTag || ((t, ts) => ts.filter(isTagDiff(t)));
+
+    const instance = useRef({
+      keyupOnDocumentListener(e) {
+        if (e.key === 'Backspace' || e.key === 'Delete') {
+          instance.removeTagsInSelection();
+          setSelection([]);
+          setPosition(null);
+        }
+        if (e.key === 'Escape') {
+          e.nativeEvent.stopImmediatePropagation();
```

Comment:
Need to use nativeEvent ?

Reply:
https://gist.github.com/ggregoire/ce7bc946212920c0a6bad8125567d001